Pattern Matching

lavirlifiliol opened the pull request:
Implementation of pattern matching, in the builtin case function.

lavirlifiliol committed:
pattern matching

Test cases not yet complete

lavirlifiliol committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	gurklang/prelude.py

decorator-factory commented on gurklang/builtin_values.py:
```diff
@@ -103,6 +103,89 @@ def close(stack: T[V, T[V, S]], scope: Scope, fail: Fail):
+
+@module.register()
+def case(stack: T[V, S], scope: Scope, fail: Fail):
+    sentinel = Atom(object())  # type: ignore
```
This would increment the atom count every time, and the traceback would look pretty mysterious if something goes wrong.
Maybe call it something impossible to create, like `{begin-case}`?

decorator-factory commented on gurklang/builtin_values.py:
```diff
@@ -103,6 +103,89 @@ def close(stack: T[V, T[V, S]], scope: Scope, fail: Fail):
+            fail(f'duplicate variable name in pattern: {variables.keys() & new_vars.keys()!r}')
+        captures.extend(stack_slots)
+        variables.update(new_vars)
+    print(captures)
```
Not sure if this side effect is intended :-)

decorator-factory commented on gurklang/builtin_values.py:
```diff
@@ -103,6 +103,89 @@ def close(stack: T[V, T[V, S]], scope: Scope, fail: Fail):
+        matched, new_stack, new_variables = _matches(pattern, rest, fail)
+        if matched:
+            return vm.call(new_stack, scope.with_members(new_variables), action)
+    return stack, scope
```
Shouldn't we `fail` when no pattern is matched?

decorator-factory commented on gurklang/builtin_values.py:
```diff
@@ -103,6 +103,89 @@ def close(stack: T[V, T[V, S]], scope: Scope, fail: Fail):
+            fail(f'a pattern must be a vector, not {pattern!r}')
+        matched, new_stack, new_variables = _matches(pattern, rest, fail)
+        if matched:
+            return vm.call(new_stack, scope.with_members(new_variables), action)
```
`vm.call` shouldn't be called from inside native functions, because this kills the stacklessness and will lead to recursion errors. 

We'll have to do something like this:
```py
@module.register()
def __get_case(stack: T[V, S], scope: Scope, fail: Fail):
    ...
    def fn(stack, scope):
        return (action, new_stack), scope.with_members(new_variables)
    return NativeFunction(fn)
    
# { get-case ! } parent-scope :case jar
module.add(
    "case",
    Code([CallByName("--get-case"), CallByValue()], name="case", flags=CodeFlags.PARENT_SCOPE)
)
```

decorator-factory committed:
Merge branch 'master' into master

lavirlifiliol committed:
fix pattern matching

stop using the python call stack
insert variables into the function scope rather than leak them
add tests for tuple pattern matching
merge new version

lavirlifiliol committed:
Merge branch 'master' of https://github.com/lavirlifiliol/gurklang-pyimpl-interpreter

lavirlifiliol committed:
add pattern matching example

lavirlifiliol committed:
implement wildcard pattern

decorator-factory commented on gurklang/prelude.py:
```diff
@@ -292,7 +315,8 @@ def name_getter(stack: Stack, scope: Scope):
         return (function, rest), scope
-    return Code([Put(NativeFunction(name_getter, name)),  CallByValue()], None, CodeFlags.PARENT_SCOPE)
+
+    return Code([Put(NativeFunction(name_getter, name)), CallByValue()], None, CodeFlags.PARENT_SCOPE)
```
There's a `raw_function` helper now in `builtin_utils`, maybe use it instead?

decorator-factory reviewed (approved):
LGTM
